# EDA of eBay Used Car Data

This is a scraped dataset of used cars from Germanys eBay. This project was meant to practice data cleaning, as well as some basic EDA analysis.

### First we import relevant packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### Then, lets read in our data

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

The first few observations:

> The dataset contains 20 columns, most of which are strings.

>Some columns have null values, but none have more than ~20% null values.

>The column names use camelcase instead of Python's preferred snakecase


### Cleaning our columns

Lets clean our columns to make them easier to work with, and read. We will:

> Convert them to snake case

> Word them more efficently if possible

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'date_created', 'num_pics', 'postal_code',
       'last_seen']

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage         date_created  num_pics  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00         0        79588   
1         bmw              nein  2016-04-04 00:00:00         0        71034   
2  volkswagen              nein  2016-03-26 00:00:00         0        35394   
3       smart              nein  2016-03-12 00:00:00         0        33729   
4        ford              nein  2016-04-01 00:00:00         0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

### Some basic data exploration to see what other cleaning we need

In [8]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

               date_created  num_pics   postal_code            last_seen  
count                 50000   50000.0  50000.000000                50000  
unique                   76       NaN           NaN                39481  
top     2016-04-03 00:00:00       NaN           NaN  2016-04-07 06:17:27  
freq                   1946       NaN           NaN                    8  
mean                    NaN       0.0  50813.627300                  NaN  
std                     NaN       0.0  25779.747957                  NaN  
min                     NaN       0.0   1067.000000                  NaN  
25%                     NaN       0.0  30451.000000                  NaN  
50%                     NaN       0.0  49577.000000                  NaN  
75%                     NaN       0.0  71540.000000                  NaN  
max                     NaN       0.0  99998.000000                  NaN

The seller and offer_type both only have one value, so its not much use. The num_photos column as well looks strange. Lets look at the num_pics column.

In [9]:
autos['num_pics'].value_counts()

0    50000
Name: num_pics, dtype: int64

So, num_pics only has 0's. So we are going to drop:

>`num_pics`

>`seller`

>`offer_type`

In [10]:
autos = autos.drop(['seller' , 'offer_type', 'num_pics'], axis = 1)

I also noticed a few columns need to be converted to a numeric type, so lets do that as well. We will convert both price and odometer.

In [11]:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(int)

In [12]:
autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',','').astype(int)

In [13]:
autos.rename({'odometer':'odometer_km'}, axis = 1, inplace = True)

In [14]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage         date_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

### Looking for outliers

In [15]:
price = autos['price']
od = autos['odometer_km']

In [16]:
print(price.value_counts().sort_index(ascending = True))

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64


We are seeing prices on the super high side, as well as a lot of values on the low side (at 0 dollars)

My idea is to remove anything listed at 0 dollars. Since its a bidding site, a bid could realistically start at 1 dollar. We need to look at the higher listed prices though to decide what to do

In [17]:
print(price.value_counts().sort_index(ascending = False).head(20))

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


So its relativley easy to see here, that there is a massive jump after 350000 to 1 million dollars. Although this could theoretically happen, it seems unreasonable. We can also see some silly pricings like 12345678 etc. We will remove anything over 1 million, and any 0's

In [18]:
autos = autos[(autos['price'] <= 350000) & (autos['price'] >= 1)]

In [19]:
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

We removed the price outliers, and now its time to move on to the odometer column

In [116]:
od.describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [117]:
od.value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

Our mean mileage for all cars is 125,000 miles roughly. Whats interesting to note though is that cars with 150,000 miles make up 65% of the total cars in our dataset

### Handling the date columns

Lets first look at all of the dates the crawler collected

In [21]:
date_crawled = autos['date_crawled'].str[:10]

In [22]:
value_dc = date_crawled.value_counts(normalize = True, dropna = False).sort_index()
value_dc

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [23]:
value_dc.describe()

count    34.000000
mean      0.029412
std       0.009762
min       0.001400
25%       0.029980
50%       0.032781
75%       0.034840
max       0.038608
Name: date_crawled, dtype: float64

The date crawled data consists of data collected daily, for around a month

Now lets check out when the posts (or ads) were created for the cars

In [24]:
date_created = autos['date_created'].str[:7]

In [25]:
date_created.value_counts(normalize = True, dropna = False).sort_index()

2015-06    0.000021
2015-08    0.000021
2015-09    0.000021
2015-11    0.000021
2015-12    0.000041
2016-01    0.000247
2016-02    0.001256
2016-03    0.837496
2016-04    0.160877
Name: date_created, dtype: float64

There were a ton of date created values, so I changed the slicing up from the full date, to the year and month. We can see most of the postings were created in March 2016 - April 2016, right around when the crawler started collecting data.

In [26]:
date_created.value_counts().sort_index()

2015-06        1
2015-08        1
2015-09        1
2015-11        1
2015-12        2
2016-01       12
2016-02       61
2016-03    40673
2016-04     7813
Name: date_created, dtype: int64

I just wanted to see the counts here to have another visualization of how these ad dates were distributed. We can see there are a few old posts, but most have been created in the year 2016

And finally lets look at the last seen column, which gives us an idea of a sale. If the post is no longer seen, than we can assume it was removed.

In [27]:
last_seen = autos['last_seen'].str[:10]

In [28]:
last_seen_configured = last_seen.value_counts(normalize = True, dropna = False).sort_index()

In [29]:
last_seen_configured.describe()

count    34.000000
mean      0.029412
std       0.043718
min       0.001071
25%       0.012432
50%       0.019489
75%       0.023783
max       0.221806
Name: last_seen, dtype: float64

We see some interesting things here. Our mean is around 3% but our min is .1% and max is 22%. I think we should look at the range here and see whats going on.

In [30]:
last_seen_configured

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [31]:
last_seen_configured = last_seen_configured[3:-3]
last_seen_configured.describe()

count    28.000000
mean      0.018239
std       0.005922
min       0.007351
25%       0.014887
50%       0.019489
75%       0.023041
max       0.028086
Name: last_seen, dtype: float64

Our last seen data has an interesting pattern. When the crawler is first collecting data, we see a very low last seen, and when the crawler is ending we see a very high amount of last seen. After some thought about how the crawler is operating, ill explain why this isnt an increase in sales, or a low amount of sales, this is just the nature of the crawler.

The last_seen column acts as a filter. We collect new data everyday and assign the last seen date as the current day we collected it (since its the first point of collection). Then the next day when we check the data to see if they still exist. If they dont, then the last seen stays the same, and if they do the last seen gets updated to the current day. 

So as our data progresses in mass, little pieces are getting left behind at the dates they were last seen. After we remove the first few days and the last few days, we see a meaningfull pattern of about 1.8% of cars being sold per day.

In [58]:
last_seen_configured = last_seen.value_counts().sort_index()[3:-3]
last_seen_configured

2016-03-08     360
2016-03-09     466
2016-03-10     518
2016-03-11     601
2016-03-12    1155
2016-03-13     432
2016-03-14     612
2016-03-15     771
2016-03-16     799
2016-03-17    1364
2016-03-18     357
2016-03-19     769
2016-03-20    1003
2016-03-21    1002
2016-03-22    1038
2016-03-23     900
2016-03-24     960
2016-03-25     933
2016-03-26     816
2016-03-27     760
2016-03-28    1013
2016-03-29    1085
2016-03-30    1203
2016-03-31    1155
2016-04-01    1107
2016-04-02    1210
2016-04-03    1224
2016-04-04    1189
Name: last_seen, dtype: int64

I wanted to see the actual counts to quantify 1.8% - Its about 900 cars

In [32]:
registration_year = autos['registration_year']
registration_year.describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [33]:
registration_year.value_counts()

2000    3156
2005    2936
1999    2897
2004    2703
2003    2699
        ... 
9000       1
1948       1
1000       1
1931       1
1952       1
Name: registration_year, Length: 95, dtype: int64

So the registration years are a bit weird. On the tail end of our values, we can see years like 1000, or 9000 which cant be right. I think im going to sort out any cars that are below 1908 (a quick google search shows that the Model T was one of the first production cars, starting in 1908. Although this is silly, older cars like this **could** be posted.

In [34]:
autos = autos[(autos['registration_year'] >= 1908) & (autos['registration_year'] <= 2022)]

In [35]:
reg_year = autos['registration_year'].value_counts(normalize = True, dropna = False).sort_index()
reg_year_1900 = autos[(autos['registration_year'] >= 1910) & (autos['registration_year'] <= 1999)]
reg_year_2000 = autos[(autos['registration_year'] >= 2000) & (autos['registration_year'] <= 2016)]



In [36]:
reg_year_1900['registration_year'].value_counts().sum()

13218

In [37]:
reg_year_2000['registration_year'].value_counts().sum()

33463

Above I created 2 new dataframes that sorted cars made in the 1900's and cars made in the 2000's because I thought it might be useful. We can see that cars made in 2000 and after are most of the postings on eBay(They account for roughly 66% of total postings). Although older cars are very popular as well.

In [38]:
reg_year_1900['registration_year'].value_counts().sort_values(ascending = False)

1999    2897
1998    2363
1997    1951
1996    1373
1995    1227
        ... 
1938       1
1927       1
1943       1
1931       1
1952       1
Name: registration_year, Length: 61, dtype: int64

There are actually more cars listed as the years get more recent which is interesting.

In [39]:
reg_year_2000['registration_year'].value_counts().sort_values(ascending = False)

2000    3156
2005    2936
2004    2703
2003    2699
2006    2670
2001    2636
2002    2486
2007    2277
2008    2215
2009    2085
2011    1623
2010    1589
2012    1310
2016    1220
2013     803
2014     663
2015     392
Name: registration_year, dtype: int64

The cars seem to peak in popularity for the year 2000, and then show a steady decline. The incline to 2000 and then the decline after should show a bell curve of sorts on a histogram.

Ive gotten a little messy with the code thusfar, so im going to clean up our original dataframe and then move forward. To do this I will remove anything outside 1908-2016 in our original dataset.

In [40]:
autos = reg_year_1900.append(reg_year_2000)

There are some cleaner ways to do what ive done so far, but since I know the dataframes match, I can easily append them together again.

### Looking into brands

In [70]:
unique_brands_normalized = autos['brand'].value_counts(normalize = True)
popular_brands_index = unique_brands_normalized[unique_brands_normalized > 0.05].index
unique_brands_normalized

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

In [68]:
popular_brands_index # Finds the brands with 5% or more frequency

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

Im only going to consider brands that have at least a 5% or more frequency in our postings. Lets just look at a basic mean price of these brands.

In [123]:
pop_brands_mean = {}
for i in popular_brands_index:
    prices = autos[autos['brand'] == i]['price'].mean()
    pop_brands_mean[i] = int(round(prices))

In [124]:
pop_brands_final = pd.Series(pop_brands_mean).sort_values(ascending = False)
pop_brands_final

audi             9337
mercedes_benz    8628
bmw              8333
volkswagen       5402
ford             3749
opel             2975
dtype: int64

Of the most popular brands, opel is the cheapest car, and audi is the most expensive. The list looks typical to what youd expect. Luxury brands are generally more expensive. I want to add the counts as well, so lets create a new dataframe so we can see the popularity and the mean price of the cars.

In [125]:
pop_brands_df = pd.DataFrame(pop_brands_final, columns = ['mean_price'])
pop_brands_df

mean_price
audi                 9337
mercedes_benz        8628
bmw                  8333
volkswagen           5402
ford                 3749
opel                 2975

In [126]:
pop_brand_counts = autos['brand'].value_counts()
pop_brands_df['counts'] = pop_brand_counts

In [127]:
pop_brands_df

mean_price  counts
audi                 9337    4041
mercedes_benz        8628    4503
bmw                  8333    5137
volkswagen           5402    9862
ford                 3749    3263
opel                 2975    5022

Now lets add the mean mileage as well for fun

In [128]:
mean_mileage = {}

for i in popular_brands_index:
    mean = autos[autos['brand'] == i]['odometer_km'].mean()
    mean_mileage[i] = int(round(mean))

In [129]:
mean_mileage

{'volkswagen': 128707,
 'bmw': 132573,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [130]:
mean_mileage = pd.Series(mean_mileage)

In [131]:
pop_brands_df['mean_mileage'] = mean_mileage

In [132]:
pop_brands_df

mean_price  counts  mean_mileage
audi                 9337    4041        129157
mercedes_benz        8628    4503        130788
bmw                  8333    5137        132573
volkswagen           5402    9862        128707
ford                 3749    3263        124266
opel                 2975    5022        129310

Im also interested in the top year for each brand (in terms of frequency), so Ill add that to our dataframe as well.

In [133]:
pop_brands_year = {}
for i in popular_brands_index:
    year = autos[autos['brand'] == i]['registration_year'].value_counts()
    top_year = year.index[0]
    pop_brands_year[i] = top_year

In [134]:
pop_brands_year = pd.Series(pop_brands_year)
pop_brands_df['most_freq_year'] = pop_brands_year
pop_brands_df

mean_price  counts  mean_mileage  most_freq_year
audi                 9337    4041        129157            2008
mercedes_benz        8628    4503        130788            2000
bmw                  8333    5137        132573            2005
volkswagen           5402    9862        128707            1999
ford                 3749    3263        124266            1999
opel                 2975    5022        129310            2000

### Conclusion

> The cars increase in popularity up until 2000, and then show a steady decline. 

> An average of 1.8%, or 900 cars, were sold everyday.

> In our final dataframe above, there seems to be three tiers of cars in terms of price. Below, I calculated the mean milage, mean price, and total popularity for these three tiers. We see come clear patterns. As price increases, so does the popularity(popularity is referring to the actual frequency in which these cars are posted), but as well as mileage. 

> This pattern stays somewhat consistent with year as well, higher tier cars tend to be newer, and declines as we decrease in price

> 65% of cars had 150,000 miles

> 66% of cars were made in the year 2000 or after

|           | Mean Price | Popularity(Total) | Mean Mileage | Most Frequent Car Year |
|-----------|------------|-------------------|--------------|------------------------|
| High Tier | 8,775      | 13,681            | 130,839      | 2000, 2005, 2008       |
| Mid Tier  | 5,402      | 9,862             | 128,707      | 1999, 2005             |
| Low Tier  | 3,362      | 8,285             | 126,788      | 1999, 2000             |